File: nodejs-soa/order-service/src/infrastructure/persistence/OrderRepository.ts
// =============================================================================
// Order Repository
// =============================================================================
// Repository Pattern for order data access.
// Handles both the orders table and order_items table as a single aggregate.
// =============================================================================

import { Pool } from 'pg';
import { v4 as uuidv4 } from 'uuid';
import { Order, OrderItem, CreateOrderData } from '../../domain/model/Order';
import { OrderStatus } from '../../domain/model/OrderStatus';

export class OrderRepository {
  constructor(private readonly pool: Pool) {}

  /**
   * Retrieve all orders with their items.
   */
  async findAll(): Promise<Order[]> {
    const ordersResult = await this.pool.query(
      'SELECT * FROM orders ORDER BY created_at DESC'
    );

    const orders: Order[] = [];
    for (const row of ordersResult.rows) {
      const items = await this.findItemsByOrderId(row.id);
      orders.push(this.mapToOrder(row, items));
    }

    return orders;
  }

  /**
   * Find an order by ID with its items.
   */
  async findById(id: string): Promise<Order | null> {
    const result = await this.pool.query('SELECT * FROM orders WHERE id = $1', [id]);
    if (result.rows.length === 0) return null;

    const items = await this.findItemsByOrderId(id);
    return this.mapToOrder(result.rows[0], items);
  }

  /**
   * Find all orders for a specific user.
   */
  async findByUserId(userId: string): Promise<Order[]> {
    const ordersResult = await this.pool.query(
      'SELECT * FROM orders WHERE user_id = $1 ORDER BY created_at DESC',
      [userId]
    );

    const orders: Order[] = [];
    for (const row of ordersResult.rows) {
      const items = await this.findItemsByOrderId(row.id);
      orders.push(this.mapToOrder(row, items));
    }

    return orders;
  }

  /**
   * Create a new order with items in a transaction.
   * Uses database transaction to ensure atomicity.
   */
  async create(data: CreateOrderData): Promise<Order> {
    const client = await this.pool.connect();

    try {
      await client.query('BEGIN');

      const orderId = uuidv4();
      const totalAmount = data.items.reduce(
        (sum, item) => sum + item.quantity * item.unitPrice,
        0
      );

      // Insert order
      const orderResult = await client.query(
        `INSERT INTO orders (id, user_id, status, total_amount)
         VALUES ($1, $2, $3, $4)
         RETURNING *`,
        [orderId, data.userId, OrderStatus.PENDING, totalAmount]
      );

      // Insert order items
      const items: OrderItem[] = [];
      for (const item of data.items) {
        const itemId = uuidv4();
        const subtotal = item.quantity * item.unitPrice;

        const itemResult = await client.query(
          `INSERT INTO order_items (id, order_id, product_id, product_name, quantity, unit_price, subtotal)
           VALUES ($1, $2, $3, $4, $5, $6, $7)
           RETURNING *`,
          [itemId, orderId, item.productId, item.productName, item.quantity, item.unitPrice, subtotal]
        );

        items.push(this.mapToOrderItem(itemResult.rows[0]));
      }

      await client.query('COMMIT');

      return this.mapToOrder(orderResult.rows[0], items);
    } catch (error) {
      await client.query('ROLLBACK');
      throw error;
    } finally {
      client.release();
    }
  }

  /**
   * Update order status.
   */
  async updateStatus(id: string, status: OrderStatus): Promise<Order | null> {
    const result = await this.pool.query(
      `UPDATE orders SET status = $1, updated_at = CURRENT_TIMESTAMP WHERE id = $2 RETURNING *`,
      [status, id]
    );

    if (result.rows.length === 0) return null;

    const items = await this.findItemsByOrderId(id);
    return this.mapToOrder(result.rows[0], items);
  }

  /**
   * Find order items by order ID.
   */
  private async findItemsByOrderId(orderId: string): Promise<OrderItem[]> {
    const result = await this.pool.query(
      'SELECT * FROM order_items WHERE order_id = $1',
      [orderId]
    );
    return result.rows.map(this.mapToOrderItem);
  }

  private mapToOrder(row: any, items: OrderItem[]): Order {
    return {
      id: row.id,
      userId: row.user_id,
      status: row.status as OrderStatus,
      totalAmount: parseFloat(row.total_amount),
      items,
      createdAt: row.created_at,
      updatedAt: row.updated_at,
    };
  }

  private mapToOrderItem(row: any): OrderItem {
    return {
      id: row.id,
      orderId: row.order_id,
      productId: row.product_id,
      productName: row.product_name,
      quantity: row.quantity,
      unitPrice: parseFloat(row.unit_price),
      subtotal: parseFloat(row.subtotal),
    };
  }
}

File: python-soa/order-service/app/infrastructure/persistence/order_repository.py
"""
Order Repository
=================
Data access layer for Order aggregates.

Design Pattern: Repository with Aggregate Loading
- Loads the complete Order aggregate (Order + OrderItems)
- Maintains consistency of the aggregate boundary
- Uses transactions for multi-table operations
"""

import logging
from decimal import Decimal
from uuid import UUID

import asyncpg

from app.domain.model import Order, OrderItem, OrderStatus

logger = logging.getLogger(__name__)


class OrderRepository:
    """
    Repository for Order persistence operations.

    Handles the Order aggregate which spans two tables:
    'orders' and 'order_items'. Operations use transactions
    to maintain consistency.
    """

    def __init__(self, pool: asyncpg.Pool):
        self._pool = pool

    async def save(self, order: Order) -> None:
        """
        Persist a new order with all its items in a transaction.

        Uses a database transaction to ensure atomicity:
        either the entire order (with items) is saved, or nothing is.
        """
        async with self._pool.acquire() as conn:
            async with conn.transaction():
                # Insert order
                await conn.execute(
                    """
                    INSERT INTO orders (id, user_id, status, total_amount, created_at, updated_at)
                    VALUES ($1, $2, $3, $4, $5, $6)
                    """,
                    order.id,
                    order.user_id,
                    order.status.value,
                    order.total_amount,
                    order.created_at,
                    order.updated_at,
                )

                # Insert order items
                for item in order.items:
                    await conn.execute(
                        """
                        INSERT INTO order_items (id, order_id, product_id, product_name, quantity, unit_price, subtotal)
                        VALUES ($1, $2, $3, $4, $5, $6, $7)
                        """,
                        item.id,
                        item.order_id,
                        item.product_id,
                        item.product_name,
                        item.quantity,
                        item.unit_price,
                        item.subtotal,
                    )

    async def find_by_id(self, order_id: UUID) -> Order | None:
        """
        Find an order by ID, loading the complete aggregate.

        Loads both the order header and all associated items.
        """
        async with self._pool.acquire() as conn:
            # Load order
            order_row = await conn.fetchrow(
                "SELECT * FROM orders WHERE id = $1", order_id
            )
            if not order_row:
                return None

            # Load items
            item_rows = await conn.fetch(
                "SELECT * FROM order_items WHERE order_id = $1", order_id
            )

            return self._map_aggregate(order_row, item_rows)

    async def find_by_user(self, user_id: UUID) -> list[Order]:
        """Find all orders for a specific user."""
        async with self._pool.acquire() as conn:
            order_rows = await conn.fetch(
                "SELECT * FROM orders WHERE user_id = $1 ORDER BY created_at DESC",
                user_id,
            )

            orders = []
            for order_row in order_rows:
                item_rows = await conn.fetch(
                    "SELECT * FROM order_items WHERE order_id = $1",
                    order_row["id"],
                )
                orders.append(self._map_aggregate(order_row, item_rows))

            return orders

    async def update_status(self, order_id: UUID, status: OrderStatus) -> None:
        """Update only the order status (used during saga transitions)."""
        async with self._pool.acquire() as conn:
            await conn.execute(
                """
                UPDATE orders SET status = $2, updated_at = CURRENT_TIMESTAMP
                WHERE id = $1
                """,
                order_id,
                status.value,
            )

    @staticmethod
    def _map_aggregate(
        order_row: asyncpg.Record, item_rows: list[asyncpg.Record]
    ) -> Order:
        """Map database rows to an Order aggregate."""
        items = [
            OrderItem(
                id=row["id"],
                order_id=row["order_id"],
                product_id=row["product_id"],
                product_name=row["product_name"] or "",
                quantity=row["quantity"],
                unit_price=Decimal(str(row["unit_price"])),
                subtotal=Decimal(str(row["subtotal"])),
            )
            for row in item_rows
        ]

        return Order(
            id=order_row["id"],
            user_id=order_row["user_id"],
            status=OrderStatus(order_row["status"]),
            total_amount=Decimal(str(order_row["total_amount"])) if order_row["total_amount"] else Decimal("0"),
            items=items,
            created_at=order_row["created_at"],
            updated_at=order_row["updated_at"],
        )
